flyrs/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+
+// Failed at encoding responded object to bytes string
+struct MessageEncodeError(EncodeError);
+// Failed at decoding requested object from bytes string
+struct MessageDecodeError(DecodeError);
+
+impl fmt::Display for MessageEncodeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "")
+    }
+}
+
+impl fmt::Display for MessageDecodeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "")
+    }
+}
+
+impl std::convert::From<MessageEncodeError> for PyErr {
+    fn from(err: MessageEncodeError) -> PyErr {
+        PyOSError::new_err(err.to_string())
+    }
+}
+
+impl std::convert::From<MessageDecodeError> for PyErr {
+    fn from(err: MessageDecodeError) -> PyErr {
+        PyOSError::new_err(err.to_string())
+    }
+}
+
+impl std::convert::From<EncodeError> for MessageEncodeError {
+    fn from(other: EncodeError) -> Self {
+        Self(other)
+    }
+}
+
+impl std::convert::From<DecodeError> for MessageDecodeError {
+    fn from(other: DecodeError) -> Self {
+        Self(other)
+    }
+}
+
+/// A Python class constructs the gRPC service stubs and a Tokio asynchronous runtime in Rust.
+#[pyclass(subclass)]
+pub struct FlyteClient {
+    admin_service: AdminServiceClient<Channel>,
+    runtime: Runtime,
+}
+
+pub fn decode_proto<T>(bytes_obj: &PyBytes) -> Result<T, MessageDecodeError>
+    where
+    T: Message + Default, {
+    let bytes = bytes_obj.as_bytes();
+    let de = Message::decode(&bytes.to_vec()[..]);
+    Ok(de?)
+}
+
+pub fn encode_proto<T>(res:T) -> Result<Vec<u8>, MessageEncodeError>
+where
+T: Message + Default, {
+    let mut buf = vec![];
+    res.encode(&mut buf)?;
+    Ok(buf)
+}
+
+#[pymethods]
+impl FlyteClient {
+    #[new] // Without this, you cannot construct the underlying class in Python.
+    #[pyo3(signature = (endpoint, **kwargs))]
+    pub fn new(endpoint: &str, kwargs: Option<&Bound<'_, PyDict>>) -> PyResult<FlyteClient> {
+        // Use Atomic Reference Counting abstractions as a cheap way to pass string reference into another thread that outlives the scope.
+        let s = Arc::new(endpoint);
+        // Check details for constructing Tokio asynchronous `runtime`: https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.new_current_thread
+        let rt = match Builder::new_current_thread().enable_all().build() {
+            Ok(rt) => rt,
+            Err(error) => panic!("Failed to initiate Tokio multi-thread runtime: {:?}", error),
+        };
+        // Check details for constructing `channel`: https://docs.rs/tonic/latest/tonic/transport/struct.Channel.html#method.builder
+        // TODO: generally handle more protocols, e.g., `https://`
+        let endpoint_uri = match format!("http://{}", *s.clone()).parse::<Uri>() {
+            Ok(uri) => uri,
+            Err(error) => panic!(
```

Comment:
Are the panics intended?

Reply:
I don't think so; I'm on my way to removing it.